### DISCUSSION TITLE
Stuff that needs to be upstreamed / fixed in `libc` and Rust STD

### DISCUSSION THREAD

**Discussion by ivmarkov** [esp-rs/esp-idf-sys#306] [Ideas]
(work in progress)

**Comment by ivmarkov** [comment]
#### [This](https://github.com/rust-lang/libc/pull/3345) PR needs to be brought to completion (for `target_os = "espidf"` at least)

The author was kind enough to provide a script that (as far as I recall) compares the sizes and values of various structs and constants hard-coded in the `libc` crate against the "real thing" (headers for the relevant platform).

So what's necessary here is to first of all adjust the script so that it uses the ESP IDF toolchain, and the ESP IDF headers. Then once it spits out its analysis, we need to fix all structs and constants reported as wrong in the `libc` crate (for ESP IDF only, of course)

**Reply by SergioGasquez** [reply]
Just opened https://github.com/rust-lang/libc/pull/3920 which should solve this

**Reply by ivmarkov** [reply]
This is now resolved, thanks to @SergioGasquez

**Comment by ivmarkov** [comment]
#### (Might be fixed by the previous PR but needs to be researched anyway)

IPv4 multicast is currently broken for the ESP IDF in Rust STD.
Here's a [relevant workaround](https://github.com/ivmarkov/esp-idf-matter/blob/master/src/multicast.rs), which uses structs and constants defined by `esp-idf-sys` rather than the ones in `libc` and thus works

**Reply by SergioGasquez** [reply]
I suppose this should be also fixed once https://github.com/rust-lang/libc/pull/3920 is merged

**Reply by ivmarkov** [reply]
This is now also resolved.

**Comment by ivmarkov** [comment]
#### Flip (or completely remove) the `espidf_time64` flag

* If we decide to completely remove it - like now - we have to immediately announce we are effectively killing any support for ESP IDF 4.4.x (reason: every ESP IDF version uses its own toolchain; the toolchain downloaded when ESP IDF 4.4.x is used has the `timeval` thing still set to 32 bit). And we'll of cpourse default to 64bit, which is compatible with ESP IDF >= 5

* If we decide to keep it, here's my suggestion (open for comments):
  - introduce a new flag: `espidf_time32`
  - keep the existing flag too: `espidf_time64`
 
Resolve `timeval` to 32 (= ESP-IDF 4) or 64 (= ESP-IDF 5) bits as follows:
```
-----------------------------------------------------------
| timeval     | espidf_time64  | espidf_time32  
|--------------------------------------------------------- 
| = 64 bits   | Defined        | Not defined     
| = 32 bits   | Not defined    | Defined           
| = 64 bits   | Not defined    | Not defined     
| Comp Err    | Defined        | Defined            
------------------------------------------------------------
```

**Reply by Vollbrecht** [reply]
In general i would like to remove the time64 flag because than it would potentially to be there forever, Some other related work that is currently done on libc is [this](https://github.com/rust-lang/libc/pull/3175) PR. I think just not checking the flag in libc is not a breaking change. A potential problem is that a person using esp-idf v4 would silently get a wrong type size. 

Should we need a time32 flag: 
- cons: it introduces minimal "dead code" for a esp-idf version we no longer support
- pro's make sure that even with a new compiler a old project can still be build.

First i thought that it might not be such a big problem to just let people stay on older nightly, and we can just ignore the introduction of a time32 flag. The one problem i have to check is "how long are older nightly" available. Situation on the riscv site; you need to recompile your old project -> install the old compiler -> because we "need" nightly need old nightly version. For a stable compiler that is not a problem but for nightly that might be a problem. That may be a trap. That would not be a problem on xtensa.

So if the above is an actual problem we **need** an time32 provided by ourself if no alternative approach is introduced into libc.

**Reply by ivmarkov** [reply]
> In general i would like to remove the time64 flag because than it would potentially to be there forever, Some other related work that is currently done on libc is [this](https://github.com/rust-lang/libc/pull/3175) PR. 

So what if it stays there forever? We are talking about 3 lines of code, after all. It is not our biggest maintenance burden, so to say.

As for the PR you are refering to - if it ever gets merged, we can consider changing our flag to this PR (given that it has enough "platform neutral" feel to it; I can't escape the feeling it is somehow hard-wired for glibc only).

But I'm not sure we should wait for it. Our flag is already in there, a change to it would be easily accepted. As it is specific to the `espidf` target platform anyway.

> I think just not checking the flag in libc is not a breaking change. 

What you mean? _The only non-breaking change is if we don't do anything (i.e., no change)_. Otherwise:
* Remove the flag completely => we lose support for ESP IDF 4.4 immediately. This is a breaking change. Unless the user uses an old nightly to compile their stuff (and potentially sits on old `esp-idf-*` crates in case we decide to raise the MSRV in the crates)
* Flip the flag meaning as I suggested => we need to inform the users that with nightlies >= X.Y.Z they need to use the `espidf_time32` flag for compiling against ESP IDF 4. Also a breaking change.

> A potential problem is that a person using esp-idf v4 would silently get a wrong type size. 

I think this is pretty much completely intorelable, but no users won't get the wrong type size, as we have a compile time error (in esp-idf-sys) if `libc` reports `timeval` of a different bit length compared to what `esp-idf-sys` reports w.r.t. `timeval`. And this compile-time check does not rely on the existence of the flag _at all_. And you can't really use `libc` without compiling `esp-idf-sys` anyway, as you don't have the ESP IDF static libs to link against.

> Should we need a time32 flag: 
> - cons: it introduces minimal "dead code" for a esp-idf version we no longer support
> - pro's make sure that even with a new compiler a old project can still be build.
> First i thought that it might not be such a big problem to just let people stay on older nightly, and we can just ignore the introduction of a time32 flag. The one problem i have to check is "how long are older nightly" available. Situation on the riscv site; you need to recompile your old project -> install the old compiler -> because we "need" nightly need old nightly version. For a stable compiler that is not a problem but for nightly that might be a problem. That may be a trap. That would not be a problem on xtensa.
> So if the above is an actual problem we **need** an time32 provided by ourself if no alternative approach is introduced into libc. 

As per above, I think you are counting too much on the `glibc work, which is ongoing for years. :)
In the end it is really simple, just two options:
* Should we support 4.4 for extra 6-9 months, with recent `esp-idf-*` crates + recent rustc = flip the flag or continue threading water
* Should we not (and then tell them to use old nightly whatever - this is the same as killing 4.4 support in my mind) = kill the flag (or continue threading water)

**Reply by Vollbrecht** [reply]
My proposal:

From a timeline perspective it would look like:
upcoming release
        :: deprecation notice v4
        :: mention in the future v4 api will be deleted (in CHANGELOG.md) 
        :: mention rustc will lose support on v4 in the same future (in CHANGELOG.md) 
"3-6 month later"in the vicinity of the future release down below
       :: adding time32 flag ( because we are nice)
       :: either keeping or removing the time64 flag
       :: setting the default if neither time64 or time32 is given to 64bit
future release
      :: removing v4 stuff

**Reply by Vollbrecht** [reply]
This way we give notice, and everybody some time to transition. And a little escape hatch with the time32 for people if they really need to.

**Reply by SergioGasquez** [reply]
In case you missed it, https://github.com/rust-lang/libc/pull/3175 seems to be progressing in the latest weeks. That being said, I like @Vollbrecht proposal.

**Reply by ivmarkov** [reply]
> In case you missed it, [rust-lang/libc#3175](https://github.com/rust-lang/libc/pull/3175) seems to be progressing in the latest weeks. That being said, I like @Vollbrecht proposal.

This issue is about GNU libc. How is that relevant to our case (ESP IDF is based on newlib)?

**Reply by SergioGasquez** [reply]
This should be resolved when https://github.com/rust-lang/libc/pull/3993 is merged

**Reply by SergioGasquez** [reply]
https://github.com/rust-lang/libc/pull/3993 got merged! Now we should wait for a `libc` release and that release to be used in `rust-lang/rust`

**Reply by SergioGasquez** [reply]
This should now be resolved!

**Comment by ivmarkov** [comment]
#### (Finally) fix the (non-working) setting of stack size via `std::thread::Builder`

This is not working since forever, but I'm kind of tired telling the users that it does not work and they have to use the _other_ workaround, so why not fixing it?

Not a high priority though.

**Comment by ivmarkov** [comment]
#### Watch out the development [here](https://github.com/rust-lang/rust/pull/125271#issuecomment-2119229310)

I think we have `posix_memalign` in ESP IDF >= 4.4, but still good to double-check, or else that would break everything for us.

**Reply by SergioGasquez** [reply]
Is there any action items on our side now that the mentioned PR has been merged?

**Reply by ivmarkov** [reply]
No this is no longer an issue. Topic closed.

**Comment by ivmarkov** [comment]
#### Talk to the LLVM Espressif team about the `async-io` [issue](https://github.com/smol-rs/async-lock/issues/84)

... and if necessary, create a small(er) test case etc.

Note: While not strictly a Rust STD / `libc` work, in a way `async-io` is _the_ way to do async IO with Rust STD for ESP-IDF.

**Comment by ivmarkov** [comment]
#### After all of the above is done - lobby for moving the riscv `-espidf` targets to Tier 2

... because the only unstable feature we use from Rust is actually in cargo, and it is `-Zbuild-std`...

So Tier 2 would give us Stable Rust for STD!

**Reply by Vollbrecht** [reply]
The build-std is beeing looked into currently with the new forming of the wg-embedded group on rust,

**Reply by ivmarkov** [reply]
Great, but still might take years until stabilized. And I fear they might stabilize only the "build-core" aspect of `build-std` initially, which is of no use to us.

Do you have a link btw?

**Reply by Vollbrecht** [reply]
There are no actionable items yet, they were discussed at the RustNL unconference between members of the embedded group and some people of rust-lang. Here is the link to what was all [discussed](https://hackmd.io/p6bQ0OzqTj-R9ra11LvP8A#Friday-PM2-Code-size)

**Reply by ivmarkov** [reply]
Yeah. Still no owner and as expected, people mainly want `core`, not the full circus. I think we should push for Tier 2 in parallel.

Unrelated: on that mtg there were complaints about async code size, but not about futures memory consumption? Weird!

**Reply by SergioGasquez** [reply]
We already started working on creating the MCP for this, we will be submitting it soon
